pymc3/distributions/multivariate.py

```diff
@@ -20,63 +20,79 @@
-        # that returns a nan as first entry instead of raising
-        # an error.
-        cholesky = Cholesky(nofail=True, lower=True)
+
+        if chol is not None and not lower:
+            chol = chol.T
+
+        self.cov = self.tau = self.chol_cov = None
 
         if cov is not None:
-            self.k = cov.shape[0]
             self._cov_type = 'cov'
-            cov = tt.as_tensor_variable(cov)
-            if cov.ndim != 2:
-                raise ValueError('cov must be two dimensional.')
-            self.chol_cov = cholesky(cov)
-            self.cov = cov
-            self._n = self.cov.shape[-1]
-        elif tau is not None:
-            self.k = tau.shape[0]
-            self._cov_type = 'tau'
-            tau = tt.as_tensor_variable(tau)
-            if tau.ndim != 2:
-                raise ValueError('tau must be two dimensional.')
-            self.chol_tau = cholesky(tau)
-            self.tau = tau
-            self._n = self.tau.shape[-1]
-        else:
-            self.k = chol.shape[0]
+            self.cov = tt.as_tensor_variable(cov)
+        elif chol is not None:
             self._cov_type = 'chol'
-            if chol.ndim != 2:
-                raise ValueError('chol must be two dimensional.')
-            self.chol_cov = tt.as_tensor_variable(chol)
-            self._n = self.chol_cov.shape[-1]
+            self.chol = tt.as_tensor_variable(chol)
+            if self.chol.ndim != 2:
+                raise ValueError('`chol` must be two dimensional.')
+        else:
+            self._cov_type = 'tau'
+            self.tau = tt.as_tensor_variable(tau)
+
+class _QuadFormBase(Continuous, _CovSet):
+    def __init__(self, mu=None, cov=None, chol=None, tau=None, lower=True,
+                 *args, **kwargs):
+        super(_QuadFormBase, self).__init__(*args, **kwargs)
+        super(_QuadFormBase, self).__initCov__(cov, tau, chol, lower)
+
+        if len(self.shape) > 2:
+            raise ValueError("Only 1 or 2 dimensions are allowed.")
+
+        _cov = getattr(self, self._cov_type)
+
+        try:
+            _deltalogp = self._logphelper(self._cov_type)
+            def deltalogp(delta):
+                return _deltalogp(_cov, delta)
+            self._delta_logp = deltalogp
+
+            _delta_logpsum = self._logpsumhelper(self._cov_type)
+            def deltalogpsum(delta):
+                return _delta_logpsum(_cov, delta)
+            self._delta_logp_sum = deltalogpsum
 
-    def _quaddist(self, value):
-        """Compute (x - mu).T @ Sigma^-1 @ (x - mu) and the logdet of Sigma."""
+        except ValueError:
+            errot = '`{}` must be two dimensional.'.format(self._cov_type)
+            raise_from(ValueError(error), None)
+
+        self.mu = tt.as_tensor_variable(mu)
+
+    def _repr_cov_params(self):
+        name = get_variable_name(getattr(self, self._cov_type))
+        return r'\mathit{{{}}}={}'.format(self._cov_type, name)
+
+    def logp(self, value, dosum=False):
```

Comment:
I don't think we should add kwargs to `logp`. Just add both `logp` and `logp_sum` to `_QuadFormBase`.

Reply:
Yes I guess I could make a helper function to check for vector values that need to be cast to nd (which is the bit I'm trying not to repeat through this kwarg)

Reply:
I don't think it would have hurt to have that check at two places, but this works as well :-)